Alfresco 6.1.1: unable to start if OOTB Alfresco metrics are disabled

**Steps to reproduce:**
Start an Alfresco 6.1.1 with Alfred Telemetry installed

**Expected behavior:** 
It works like a charm

**Actual behavior:**
Alfresco fails to start. Stacktrace:
```
 org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'alfred-telemetry.MeterBinderRegistrar' defined in class path resource [alfresco/module/alfred-telemetry-platform/context/binder-context.xml]: Invocation of init method failed; nested exception is java.lang.NullPointerException
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1778)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:593)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:515)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:845)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:877)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:549)
	at org.springframework.web.context.ContextLoader.configureAndRefreshWebApplicationContext(ContextLoader.java:400)
	at org.springframework.web.context.ContextLoader.initWebApplicationContext(ContextLoader.java:291)
	at org.springframework.web.context.ContextLoaderListener.contextInitialized(ContextLoaderListener.java:103)
	at org.alfresco.web.app.ContextLoaderListener.contextInitialized(ContextLoaderListener.java:70)
	at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4701)
	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5167)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
	at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:743)
	at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:719)
	at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:705)
	at org.apache.catalina.startup.HostConfig.deployDirectory(HostConfig.java:1125)
	at org.apache.catalina.startup.HostConfig$DeployDirectory.run(HostConfig.java:1859)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.NullPointerException
	at io.micrometer.core.instrument.Gauge$Builder.register(Gauge.java:147)
	at io.micrometer.core.instrument.composite.CompositeGauge.registerNewMeter(CompositeGauge.java:58)
	at io.micrometer.core.instrument.composite.CompositeGauge.registerNewMeter(CompositeGauge.java:27)
	at io.micrometer.core.instrument.composite.AbstractCompositeMeter.add(AbstractCompositeMeter.java:66)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at java.base/java.util.Collections$SetFromMap.forEach(Collections.java:5581)
	at io.micrometer.core.instrument.composite.CompositeMeterRegistry.lambda$null$0(CompositeMeterRegistry.java:59)
	at io.micrometer.core.instrument.composite.CompositeMeterRegistry.lock(CompositeMeterRegistry.java:176)
	at io.micrometer.core.instrument.composite.CompositeMeterRegistry.lambda$new$1(CompositeMeterRegistry.java:59)
	at io.micrometer.core.instrument.MeterRegistry.getOrCreateMeter(MeterRegistry.java:569)
	at io.micrometer.core.instrument.MeterRegistry.registerMeterIfNecessary(MeterRegistry.java:528)
	at io.micrometer.core.instrument.MeterRegistry.registerMeterIfNecessary(MeterRegistry.java:514)
	at io.micrometer.core.instrument.MeterRegistry.gauge(MeterRegistry.java:244)
	at io.micrometer.core.instrument.Gauge$Builder.register(Gauge.java:147)
	at io.micrometer.core.instrument.binder.jvm.JvmGcMetrics.bindTo(JvmGcMetrics.java:110)
	at eu.xenit.alfred.telemetry.binder.JvmMetrics.bindTo(JvmMetrics.java:15)
	at eu.xenit.alfred.telemetry.binder.MeterBinderRegistrar.lambda$addBinders$0(MeterBinderRegistrar.java:57)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at eu.xenit.alfred.telemetry.binder.MeterBinderRegistrar.addBinders(MeterBinderRegistrar.java:56)
	at eu.xenit.alfred.telemetry.binder.MeterBinderRegistrar.afterPropertiesSet(MeterBinderRegistrar.java:49)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1837)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1774)
	... 26 more

```

--- a/alfred-telemetry-platform/src/main/java/eu/xenit/alfred/telemetry/alfrescointegration/AlfredTelemetryMetricsController.java
+++ b/alfred-telemetry-platform/src/main/java/eu/xenit/alfred/telemetry/alfrescointegration/AlfredTelemetryMetricsController.java
@@ -3,13 +3,20 @@ package eu.xenit.alfred.telemetry.alfrescointegration;
 import io.micrometer.core.instrument.MeterRegistry;
 import io.micrometer.core.instrument.composite.CompositeMeterRegistry;
 import org.alfresco.enterprise.metrics.MetricsController;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * Implementation of Alfresco's {@link MetricsController} that will offer the global MicroMeter registry as registry.
  */
 public class AlfredTelemetryMetricsController implements MetricsController {
 
-    private CompositeMeterRegistry globalRegistry;
+    private static final Logger log = LoggerFactory.getLogger(AlfredTelemetryMetricsController.class);
+
+    private static final String MSG_INTEGRATION_FAILED =
+            "Failed to integrate Alfred Telemetry with the out of the box Alfresco metrics: ";
+
+    private final CompositeMeterRegistry globalRegistry;
     private final MetricsController alfrescoMetricsController;
 
     AlfredTelemetryMetricsController(CompositeMeterRegistry globalRegistry,
@@ -23,6 +30,15 @@ public class AlfredTelemetryMetricsController implements MetricsController {
 
     private void registerAlfrescoControllerInGlobalRegistry(final boolean enableDefaultAlfrescoRegistry) {
         if (enableDefaultAlfrescoRegistry) {
+            if (!alfrescoMetricsController.isEnabled()) {
+                log.warn(MSG_INTEGRATION_FAILED
+                        + "out of the box Alfresco metrics not enabled, see ${metrics.enabled} property.");
+                return;
+            }
+            if (alfrescoMetricsController.getRegistry() == null) {
+                log.warn(MSG_INTEGRATION_FAILED + "out of the box Alfresco MeterRegistry not correctly initialized.");
+                return;
+            }
             globalRegistry.add(alfrescoMetricsController.getRegistry());
         }
     }

--- a/alfred-telemetry-platform/src/main/java/eu/xenit/alfred/telemetry/alfrescointegration/AlfredTelemetryMetricsController.java
+++ b/alfred-telemetry-platform/src/main/java/eu/xenit/alfred/telemetry/alfrescointegration/AlfredTelemetryMetricsController.java
@@ -35,11 +35,12 @@ public class AlfredTelemetryMetricsController implements MetricsController {
                         + "out of the box Alfresco metrics not enabled, see ${metrics.enabled} property.");
                 return;
             }
-            if (alfrescoMetricsController.getRegistry() == null) {
+            final MeterRegistry alfrescoRegistry = alfrescoMetricsController.getRegistry();
+            if (alfrescoRegistry == null) {
                 log.warn(MSG_INTEGRATION_FAILED + "out of the box Alfresco MeterRegistry not correctly initialized.");
                 return;
             }
-            globalRegistry.add(alfrescoMetricsController.getRegistry());
+            globalRegistry.add(alfrescoRegistry);
         }
     }
 

--- a/alfred-telemetry-platform/src/main/java/eu/xenit/alfred/telemetry/alfrescointegration/MetricsControllerBeanPostProcessor.java
+++ b/alfred-telemetry-platform/src/main/java/eu/xenit/alfred/telemetry/alfrescointegration/MetricsControllerBeanPostProcessor.java
@@ -70,6 +70,8 @@ public class MetricsControllerBeanPostProcessor implements BeanDefinitionRegistr
 
         alfredMetricsController.setBeanClass(AlfredTelemetryMetricsController.class);
         alfredMetricsController.setConstructorArgumentValues(constructorArgumentValues);
+        // Workaround to make sure Alfresco's logging hierarchy is initialized before our bean so no logging is lost:
+        alfredMetricsController.setDependsOn("log4JHierarchyInit");
 
         registry.registerBeanDefinition(METRICS_CONTROLLER_BEAN_ID, alfredMetricsController);
         LOGGER.info("Bean '{}' overwritten with custom '{}' implementation", METRICS_CONTROLLER_BEAN_ID,

--- a/alfred-telemetry-platform/src/test/java/eu/xenit/alfred/telemetry/alfrescointegration/AlfredTelemetryMetricsControllerTest.java
+++ b/alfred-telemetry-platform/src/test/java/eu/xenit/alfred/telemetry/alfrescointegration/AlfredTelemetryMetricsControllerTest.java
@@ -4,6 +4,7 @@ import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.instanceOf;
 import static org.hamcrest.Matchers.is;
 import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
@@ -32,6 +33,7 @@ class AlfredTelemetryMetricsControllerTest {
     @BeforeEach
     void setup() {
         when(alfrescoMetricsController.getRegistry()).thenReturn(alfrescoDefaultRegistry);
+        when(alfrescoMetricsController.isEnabled()).thenReturn(true);
     }
 
     @Test
@@ -40,6 +42,16 @@ class AlfredTelemetryMetricsControllerTest {
 
         verify(alfrescoMetricsController).getRegistry();
         verify(meterRegistry).add(alfrescoDefaultRegistry);
+    }
+
+    @Test
+    void initialize_doesNotIncludeAlfrescoDefaultRegistryIfDisabled() {
+        when(alfrescoMetricsController.isEnabled()).thenReturn(false);
+        new AlfredTelemetryMetricsController(meterRegistry, alfrescoMetricsController, true);
+
+        verify(alfrescoMetricsController).isEnabled();
+        verify(alfrescoMetricsController, never()).getRegistry();
+        verify(meterRegistry, never()).add(alfrescoDefaultRegistry);
     }
 
     @Test
@@ -59,12 +71,10 @@ class AlfredTelemetryMetricsControllerTest {
 
     @Test
     void isEnabled_delegatesToAlfrescoDefaultRegistry() {
-        when(alfrescoMetricsController.isEnabled()).thenReturn(true);
-
         MetricsController controller = new AlfredTelemetryMetricsController(meterRegistry, alfrescoMetricsController,
                 true);
         assertThat(controller.isEnabled(), is(true));
-        verify(alfrescoMetricsController).isEnabled();
+        verify(alfrescoMetricsController, times(2)).isEnabled();
     }
 
     @Test

--- a/integration-tests/alfresco-enterprise-61/overload.gradle
+++ b/integration-tests/alfresco-enterprise-61/overload.gradle
@@ -1,4 +1,4 @@
 ext {
-    alfrescoBaseWar = 'org.alfresco:content-services:6.1.0.3@war'
-    alfrescoBaseImage = 'hub.xenit.eu/alfresco-enterprise/alfresco-repository-enterprise:6.1.0.3'
+    alfrescoBaseWar = 'org.alfresco:content-services:6.1.1@war'
+    alfrescoBaseImage = 'hub.xenit.eu/alfresco-enterprise/alfresco-repository-enterprise:6.1.1'
 }

--- a/integration-tests/src/test/java/eu/xenit/alfred/telemetry/integrationtesting/AlfrescoPrometheusEndpointTest.java
+++ b/integration-tests/src/test/java/eu/xenit/alfred/telemetry/integrationtesting/AlfrescoPrometheusEndpointTest.java
@@ -1,0 +1,35 @@
+package eu.xenit.alfred.telemetry.integrationtesting;
+
+import static io.restassured.RestAssured.given;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.isOneOf;
+
+import io.restassured.response.ExtractableResponse;
+import io.restassured.response.Response;
+import org.junit.jupiter.api.Test;
+
+public class AlfrescoPrometheusEndpointTest extends RestAssuredTest {
+
+    @Test
+    void alfrescoEndpointExposesAlfredTelemetryMetrics() {
+        ExtractableResponse<Response> response =
+                given()
+                        .log().ifValidationFails()
+                        .when()
+                        .get("s/prometheus")
+                        .then()
+                        .log().ifValidationFails()
+                        .statusCode(isOneOf(200, 404))
+                        .extract();
+
+        if (response.statusCode() == 404) {
+            return;
+        }
+
+        String responseBody = response.body().asString();
+
+        assertThat(responseBody, containsString("alfred_telemetry_registries_total"));
+    }
+
+}

--- a/integration-tests/src/test/resources/compose/docker-compose.yml
+++ b/integration-tests/src/test/resources/compose/docker-compose.yml
@@ -4,12 +4,13 @@ services:
   alfresco:
     image: ${DOCKER_IMAGE}
     ports:
-     - target: 8080
-       mode: host
+      - target: 8080
+        mode: host
     environment:
       - CATALINA_OPTS=-Xdebug -agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=8000
       - INDEX=noindex
       - ENABLE_CLUSTERING=true
+      - GLOBAL_metrics.enabled=true
       - GLOBAL_alfred.telemetry.alfresco-integration.enabled=true
       # Enable Care4Alf metrics:
       - GLOBAL_alfred.telemetry.binder.care4alf.enabled=true

